shared/templates/templates.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package templates
+
+import (
+	"bytes"
+	"html/template"
+	"io"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// Container represents a set of templates that can be render
+type Container struct {
+	templates *template.Template
+	mutex     sync.RWMutex
+	stop      chan struct{}
+	stopped   chan struct{}
+	watch     bool
+}
+
+// Data contains the data used to populate the template variables, it has Props
+// that can be of any type and HTML that only can be `template.HTML` types.
+type Data struct {
+	Props map[string]interface{}
+	HTML  map[string]template.HTML
+}
+
+// NewFromTemplates creates a new templates container using a
+// `template.Template` object
+func NewFromTemplate(templates *template.Template) *Container {
+	return &Container{templates: templates}
+}
+
+// New creates a new templates container scanning a directory.
+func New(directory string) (*Container, error) {
+	ret := &Container{}
+
+	htmlTemplates, err := template.ParseGlob(filepath.Join(directory, "*.html"))
+	if err != nil {
+		return nil, err
+	}
+	ret.templates = htmlTemplates
+
+	return ret, nil
+}
+
+// NewWithWatcher creates a new templates container scanning a directory and
+// watch the directory filesystem changes to apply them to the loaded
+// templates. The caller must consume the returned error channel to ensure not
+// blocking the watch process.
+func NewWithWatcher(directory string) (*Container, <-chan error) {
+	errors := make(chan error)
+	ret := &Container{}
+
+	htmlTemplates, err := template.ParseGlob(filepath.Join(directory, "*.html"))
+	if err != nil {
```

Comment:
I though about that, but I think is not "fair" to not return the root cause of the error. Maybe we can return `(*Container, error, chan error)`, that way, the `error` is the synchronous error about initializing the templates with a watcher, and the `chan error` is the errors produced asynchronously during the watch.

About the caller, we weren't stopping entirely the server start because the templates failure, so... My expectation would be to allow you to keep going without the htmlTemplateWatcher even if that fails to start. Also, we can use directly the `templates.NewFromTemplate(template.New(""))` on the initialization as part of the error handling, so we always have a valid `htmlTemplateWatcher` object.

Probably if I'm not able to start the `NewWithWatcher` I would log that, and try to create with a regular `New`, and finally, if that also fails, create the `templates.NewFromTemplate(template.New(""))`.

Reply:
> Maybe we can return (*Container, error, chan error), that way, the error is the synchronous error about initializing the templates with a watcher, and the chan error is the errors produced asynchronously during the watch.

Yeah .. I had thought about that. It feels a bit weird API wise to return error and chan error, as there's no clear distinction. I wanted to give a last shot at the "error chan being a struct field" idea. We would only initialize the channel in the watcher case, and send errors only from the watcher loop. And the constructor definition can remain `*Container, error`. What do you think about that?

> About the caller, we weren't stopping entirely the server start because the templates failure, 

Would there be a reason not to fail? One of our core principles is to "fail fast". I think from a robustness perspective, it's good to always start with a correct and stable state. It also removes the need to add all of those extra logic you mentioned just to have a non-nil templates object. I wanted to take this opportunity and improve some of the areas where we should be more robust, but we aren't.

Reply:
That hidden field that you have to handled... I think is more hard to follow and understand, I think the error + errorChannel is more explicit about the need of handling both.

I'm ok about "failing fast" here.

Reply:
Sounds good then, let's go with it and fail fast.

Reply:
Done